File: ios/ExpoVoicevoxModule.swift
import ExpoModulesCore

/// `initialize()` に渡される JS 側のオプション。
///
/// パスが省略された（null の）場合は、config plugin が配置したアセットを自動で解決する。
struct VoicevoxInitializeOptions: Record {
  @Field
  var openJtalkDictDir: String? = nil

  @Field
  var voiceModelPaths: [String]? = nil

  @Field
  var cpuNumThreads: Int = 0

  /// 合成結果のキャッシュに使う上限バイト数。0 で無効。
  ///
  /// Int で受けると 32bit 環境で 2GB を超える指定が無言で壊れるので、JS の number をそのまま
  /// Double で受けて `cacheLimitBytes(_:)` で検証する。
  @Field
  var synthesisCacheBytes: Double = Double(VoicevoxWavCache.defaultLimitBytes)
}

/// ユーザー辞書へ登録する単語。既定値は JS 側で埋まっている。
struct VoicevoxUserDictWordRecord: Record {
  @Field
  var surface: String = ""

  @Field
  var pronunciation: String = ""

  @Field
  var accentType: Int = 0

  @Field
  var wordType: String = "COMMON_NOUN"

  @Field
  var priority: Int = 5
}

/// voicevox-core 由来のエラーを JS へ伝えるための例外。
final class VoicevoxException: GenericException<String> {
  override var reason: String {
    param
  }
}

public class ExpoVoicevoxModule: Module {
  private let engine = VoicevoxEngine()

  /// 再生役。合成とは別に動くので直列キューには載せない。
  private let player = VoicevoxPlayer()

  /// voicevox-core の Synthesizer は同時実行できないので、重い処理は 1 本の直列キューに載せる。
  private let engineQueue = DispatchQueue(label: "expo.modules.voicevox.engine")

  /// 合成結果の LRU キャッシュ。engineQueue の上でだけ触るので、自前のロックは持たない。
  private let wavCache = VoicevoxWavCache()

  public func definition() -> ModuleDefinition {
    Name("ExpoVoicevox")

    Events("onPrepareProgress", "onSpeechStateChange")

    OnCreate {
      // モジュールが player を所有するので、コールバック側は弱参照にして循環させない。
      self.player.onStateChange = { [weak self] id, state, reason in
        self?.sendEvent(
          "onSpeechStateChange",
          ["id": id, "state": state.rawValue, "reason": reason]
        )
      }
    }

    OnDestroy {
      // 停止は engineQueue に依存しないので先に済ませる（合成の実行中でも即座に黙る）。
      self.player.stop()
      // 合成の実行中に破棄されうるので、直列キューの上で解放する。
      self.engineQueue.sync {
        self.engine.release()
        self.wavCache.clear()
      }
    }

    Function("getVersion") { () -> String in
      VoicevoxEngine.version()
    }

    Function("isInitialized") { () -> Bool in
      self.engine.isInitialized
    }

    Function("isSpeaking") { () -> Bool in
      self.player.isSpeaking
    }

    AsyncFunction("prepareAssets") { () -> [String: Any] in
      let paths = try self.prepareAssets()
      return [
        "openJtalkDictDir": paths.openJtalkDictDir,
        "voiceModelPaths": paths.voiceModelPaths,
      ]
    }
    .runOnQueue(engineQueue)

    AsyncFunction("initialize") { (options: VoicevoxInitializeOptions) in
      guard let cpuNumThreads = UInt16(exactly: options.cpuNumThreads) else {
        throw VoicevoxException("cpuNumThreads is out of range: \(options.cpuNumThreads)")
      }

      // 明示パスが両方そろっているときはアセットの準備を一切走らせない
      // （自前でモデルを管理している利用者に余計なダウンロードをさせないため）。
      let dictDir: String
      let modelPaths: [String]
      if let explicitDict = options.openJtalkDictDir, let explicitModels = options.voiceModelPaths {
        dictDir = explicitDict
        modelPaths = explicitModels
      } else {
        let prepared = try self.prepareAssets()
        dictDir = options.openJtalkDictDir ?? prepared.openJtalkDictDir
        modelPaths = options.voiceModelPaths ?? prepared.voiceModelPaths
      }

      let limitBytes = try self.cacheLimitBytes(options.synthesisCacheBytes)
      try self.wrappingErrors {
        try self.engine.initialize(
          openJtalkDictDir: dictDir,
          voiceModelPaths: modelPaths,
          cpuNumThreads: cpuNumThreads
        )
      }
      // 読み込むモデルが変わりうるので、初期化のたびに作り直す。
      self.wavCache.limitBytes = limitBytes
      self.wavCache.clear()
    }
    .runOnQueue(engineQueue)

    AsyncFunction("getMetasJson") { () -> String in
      try self.wrappingErrors { try self.engine.metasJson() }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("tts") {
      (text: String, styleId: Int, enableInterrogativeUpspeak: Bool, directory: String,
        useCache: Bool) -> String in
      let key = VoicevoxWavCache.key(
        kind: "text",
        styleId: styleId,
        enableInterrogativeUpspeak: enableInterrogativeUpspeak,
        payload: text
      )
      return try self.synthesize(
        styleId: styleId, directory: directory, cacheKey: key, useCache: useCache
      ) { styleId in
        try self.engine.tts(
          text: text,
          styleId: styleId,
          enableInterrogativeUpspeak: enableInterrogativeUpspeak
        )
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("ttsFromKana") {
      (kana: String, styleId: Int, enableInterrogativeUpspeak: Bool, directory: String,
        useCache: Bool) -> String in
      let key = VoicevoxWavCache.key(
        kind: "kana",
        styleId: styleId,
        enableInterrogativeUpspeak: enableInterrogativeUpspeak,
        payload: kana
      )
      return try self.synthesize(
        styleId: styleId, directory: directory, cacheKey: key, useCache: useCache
      ) { styleId in
        try self.engine.ttsFromKana(
          kana: kana,
          styleId: styleId,
          enableInterrogativeUpspeak: enableInterrogativeUpspeak
        )
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("createAudioQueryJson") { (text: String, styleId: Int) -> String in
      let styleId = try self.checkedStyleId(styleId)
      return try self.wrappingErrors {
        try self.engine.createAudioQueryJson(text: text, styleId: styleId)
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("createAudioQueryFromKanaJson") { (kana: String, styleId: Int) -> String in
      let styleId = try self.checkedStyleId(styleId)
      return try self.wrappingErrors {
        try self.engine.createAudioQueryFromKanaJson(kana: kana, styleId: styleId)
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("createAccentPhrasesJson") { (text: String, styleId: Int) -> String in
      let styleId = try self.checkedStyleId(styleId)
      return try self.wrappingErrors {
        try self.engine.createAccentPhrasesJson(text: text, styleId: styleId)
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("createAccentPhrasesFromKanaJson") { (kana: String, styleId: Int) -> String in
      let styleId = try self.checkedStyleId(styleId)
      return try self.wrappingErrors {
        try self.engine.createAccentPhrasesFromKanaJson(kana: kana, styleId: styleId)
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("replaceMoraDataJson") { (accentPhrasesJson: String, styleId: Int) -> String in
      let styleId = try self.checkedStyleId(styleId)
      return try self.wrappingErrors {
        try self.engine.replaceMoraDataJson(accentPhrasesJson: accentPhrasesJson, styleId: styleId)
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("replacePhonemeLengthJson") { (accentPhrasesJson: String, styleId: Int) -> String
      in
      let styleId = try self.checkedStyleId(styleId)
      return try self.wrappingErrors {
        try self.engine.replacePhonemeLengthJson(
          accentPhrasesJson: accentPhrasesJson, styleId: styleId)
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("replaceMoraPitchJson") { (accentPhrasesJson: String, styleId: Int) -> String in
      let styleId = try self.checkedStyleId(styleId)
      return try self.wrappingErrors {
        try self.engine.replaceMoraPitchJson(accentPhrasesJson: accentPhrasesJson, styleId: styleId)
      }
    }
    .runOnQueue(engineQueue)

    // 推論を伴わないので直列キューに載せる必要が無い。
    AsyncFunction("audioQueryFromAccentPhrasesJson") { (accentPhrasesJson: String) -> String in
      try self.wrappingErrors {
        try VoicevoxEngine.audioQueryFromAccentPhrasesJson(accentPhrasesJson)
      }
    }

    AsyncFunction("synthesis") {
      (audioQueryJson: String, styleId: Int, enableInterrogativeUpspeak: Bool, directory: String,
        useCache: Bool) -> String in
      let key = VoicevoxWavCache.key(
        kind: "query",
        styleId: styleId,
        enableInterrogativeUpspeak: enableInterrogativeUpspeak,
        payload: audioQueryJson
      )
      return try self.synthesize(
        styleId: styleId, directory: directory, cacheKey: key, useCache: useCache
      ) { styleId in
        try self.engine.synthesis(
          audioQueryJson: audioQueryJson,
          styleId: styleId,
          enableInterrogativeUpspeak: enableInterrogativeUpspeak
        )
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("speak") {
      (text: String, styleId: Int, enableInterrogativeUpspeak: Bool, audioSession: String,
        useCache: Bool, promise: Promise) in
      let key = VoicevoxWavCache.key(
        kind: "text",
        styleId: styleId,
        enableInterrogativeUpspeak: enableInterrogativeUpspeak,
        payload: text
      )
      self.speakWav(
        styleId: styleId, audioSession: audioSession, cacheKey: key, useCache: useCache,
        promise: promise
      ) { styleId in
        try self.engine.tts(
          text: text,
          styleId: styleId,
          enableInterrogativeUpspeak: enableInterrogativeUpspeak
        )
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("speakFromKana") {
      (kana: String, styleId: Int, enableInterrogativeUpspeak: Bool, audioSession: String,
        useCache: Bool, promise: Promise) in
      let key = VoicevoxWavCache.key(
        kind: "kana",
        styleId: styleId,
        enableInterrogativeUpspeak: enableInterrogativeUpspeak,
        payload: kana
      )
      self.speakWav(
        styleId: styleId, audioSession: audioSession, cacheKey: key, useCache: useCache,
        promise: promise
      ) { styleId in
        try self.engine.ttsFromKana(
          kana: kana,
          styleId: styleId,
          enableInterrogativeUpspeak: enableInterrogativeUpspeak
        )
      }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("speakFromAudioQuery") {
      (audioQueryJson: String, styleId: Int, enableInterrogativeUpspeak: Bool,
        audioSession: String, useCache: Bool, promise: Promise) in
      let key = VoicevoxWavCache.key(
        kind: "query",
        styleId: styleId,
        enableInterrogativeUpspeak: enableInterrogativeUpspeak,
        payload: audioQueryJson
      )
      self.speakWav(
        styleId: styleId, audioSession: audioSession, cacheKey: key, useCache: useCache,
        promise: promise
      ) { styleId in
        try self.engine.synthesis(
          audioQueryJson: audioQueryJson,
          styleId: styleId,
          enableInterrogativeUpspeak: enableInterrogativeUpspeak
        )
      }
    }
    .runOnQueue(engineQueue)

    // 合成の実行中でも即座に止められるよう、直列キューには載せない。
    AsyncFunction("stopSpeaking") {
      self.player.stop()
    }

    AsyncFunction("setUserDictWords") { (words: [VoicevoxUserDictWordRecord]) in
      let converted = try words.map { try self.toUserDictWord($0) }
      try self.wrappingErrors { try self.engine.setUserDictWords(converted) }
      // 読みが変わるので、捨てないと古い発音のまま鳴ってしまう。
      self.wavCache.clear()
    }
    .runOnQueue(engineQueue)

    AsyncFunction("loadUserDictFile") { (path: String) in
      try self.wrappingErrors { try self.engine.loadUserDictFile(path: path) }
      // setUserDictWords と同じ理由で捨てる。
      self.wavCache.clear()
    }
    .runOnQueue(engineQueue)

    AsyncFunction("saveUserDictFile") { (path: String) in
      try self.wrappingErrors { try self.engine.saveUserDictFile(path: path) }
    }
    .runOnQueue(engineQueue)

    AsyncFunction("finalize") {
      self.engine.release()
      self.wavCache.clear()
    }
    .runOnQueue(engineQueue)

    // engineQueue に載るので合成の実行中は待たされる。同期関数にすると JS スレッドが
    // 合成の完了まで止まるため、どちらも AsyncFunction のままにしておくこと。
    AsyncFunction("clearSynthesisCache") {
      self.wavCache.clear()
    }
    .runOnQueue(engineQueue)

    AsyncFunction("getSynthesisCacheStats") { () -> [String: Any] in
      self.wavCache.statsDictionary
    }
    .runOnQueue(engineQueue)
  }

  /// JS の number を上限バイト数へ直す。
  ///
  /// 検証は JS 側にもあるが、ネイティブへ直接来た値で LRU が壊れないようここでも見る。
  private func cacheLimitBytes(_ raw: Double) throws -> Int {
    guard raw.isFinite, raw >= 0, let limit = Int(exactly: raw.rounded(.down)) else {
      throw VoicevoxException("synthesisCacheBytes is out of range: \(raw)")
    }
    return limit
  }

  /// キャッシュを引いてから合成する。engineQueue の上でだけ呼ぶこと。
  ///
  /// `useCache` が false のときは読みも書きもしない。一度きりの動的なテキストで LRU を
  /// 汚さないための逃げ道なので、「読むが書かない」にはしない。
  private func cachedWav(key: String, useCache: Bool, _ body: () throws -> Data) throws -> Data {
    if useCache, let hit = wavCache.value(forKey: key) {
      return hit
    }
    let wav = try body()
    if useCache {
      wavCache.setValue(wav, forKey: key)
    }
    return wav
  }

  private func toUserDictWord(_ record: VoicevoxUserDictWordRecord) throws -> UserDictWord {
    // C の enum 定数は voicevox_core を import している VoicevoxEngine.swift 側で引く。
    guard let wordType = UserDictWord.wordType(named: record.wordType) else {
      throw VoicevoxException("unknown user dictionary word type: \(record.wordType)")
    }
    guard let accentType = UInt(exactly: record.accentType) else {
      throw VoicevoxException("accentType is out of range: \(record.accentType)")
    }
    guard let priority = UInt8(exactly: record.priority) else {
      throw VoicevoxException("priority is out of range: \(record.priority)")
    }
    return UserDictWord(
      surface: record.surface,
      pronunciation: record.pronunciation,
      accentType: accentType,
      wordType: wordType,
      priority: priority
    )
  }

  private func checkedStyleId(_ styleId: Int) throws -> UInt32 {
    guard let styleId = UInt32(exactly: styleId) else {
      throw VoicevoxException("styleId is out of range: \(styleId)")
    }
    return styleId
  }

  /// voicevox-core 由来のエラーを JS へ流せる形に包む。
  private func wrappingErrors<T>(_ body: () throws -> T) throws -> T {
    do {
      return try body()
    } catch let error as VoicevoxException {
      throw error
    } catch {
      throw VoicevoxException(error.localizedDescription)
    }
  }

  /// 合成系に共通する「styleId を検証し、WAV を指定のディレクトリへ書き出してパスを返す」流れ。
  private func synthesize(
    styleId: Int,
    directory: String,
    cacheKey: String,
    useCache: Bool,
    _ body: (UInt32) throws -> Data
  ) throws -> String {
    let styleId = try checkedStyleId(styleId)
    let searchPath = try searchPathDirectory(directory)
    return try wrappingErrors {
      let wav = try self.cachedWav(key: cacheKey, useCache: useCache) { try body(styleId) }
      return try self.writeWav(wav, to: searchPath)
    }
  }

  /// 再生系に共通する「engineQueue の上で合成し、再生はその外で始める」流れ。
  ///
  /// 再生の完了を engineQueue の中で待つと、鳴っているあいだ次の合成を始められない。
  /// `player.play` はメインキューへ投げて即座に戻り、Promise はそちらで解決する。
  private func speakWav(
    styleId: Int,
    audioSession: String,
    cacheKey: String,
    useCache: Bool,
    promise: Promise,
    _ body: (UInt32) throws -> Data
  ) {
    do {
      guard let session = VoicevoxAudioSessionMode(rawValue: audioSession) else {
        throw VoicevoxException("unknown audio session mode: \(audioSession)")
      }
      let styleId = try checkedStyleId(styleId)
      let id = player.begin()
      let wav: Data
      do {
        wav = try wrappingErrors {
          try self.cachedWav(key: cacheKey, useCache: useCache) { try body(styleId) }
        }
      } catch {
        // 合成に失敗したら自分の予約だけ畳む。鳴っている音は止めない。
        player.cancel(id: id)
        throw error
      }
      player.play(id: id, wav: wav, session: session) { result in
        switch result {
        case .success(let playback):
          promise.resolve(playback.dictionary)
        case .failure(let error):
          promise.reject(error)
        }
      }
    } catch {
      promise.reject(error)
    }
  }

  /// 書き出し先の指定を `FileManager` の検索パスへ変換する。
  private func searchPathDirectory(_ directory: String) throws -> FileManager.SearchPathDirectory {
    switch directory {
    case "cache":
      return .cachesDirectory
    // iCloud バックアップの対象。UIFileSharingEnabled 次第で Files アプリにも見える。
    case "document":
      return .documentDirectory
    default:
      throw VoicevoxException("unknown output directory: \(directory)")
    }
  }

  /// config plugin が配置したアセットを使える状態にする。進捗は JS へイベントで流す。
  private func prepareAssets() throws -> VoicevoxAssetPaths {
    do {
      return try VoicevoxAssets.shared.prepare { progress in
        self.sendEvent("onPrepareProgress", progress.dictionary)
      }
    } catch {
      throw VoicevoxException(error.localizedDescription)
    }
  }

  /// 合成結果を指定のディレクトリへ書き出し、そのパスを返す。
  private func writeWav(_ wav: Data, to searchPath: FileManager.SearchPathDirectory) throws -> String
  {
    let baseDirectory = FileManager.default.urls(for: searchPath, in: .userDomainMask)[0]
    let outputDirectory = baseDirectory.appendingPathComponent("expo-voicevox", isDirectory: true)

    do {
      try FileManager.default.createDirectory(
        at: outputDirectory,
        withIntermediateDirectories: true
      )
      let outputUrl = outputDirectory.appendingPathComponent("\(UUID().uuidString).wav")
      try wav.write(to: outputUrl, options: .atomic)
      return outputUrl.path
    } catch {
      throw VoicevoxException("failed to write the WAV file: \(error.localizedDescription)")
    }
  }
}

File: android/src/main/java/expo/modules/voicevox/ExpoVoicevoxModule.kt
package expo.modules.voicevox

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.io.File
import java.util.UUID
import jp.hiroshiba.voicevoxcore.GlobalInfo
import jp.hiroshiba.voicevoxcore.UserDictWord

/**
 * `initialize()` に渡される JS 側のオプション。
 *
 * パスが省略された（null の）場合は、config plugin が配置したアセットを自動で解決する。
 */
class VoicevoxInitializeOptions : Record {
  @Field val openJtalkDictDir: String? = null

  @Field val voiceModelPaths: List<String>? = null

  @Field val cpuNumThreads: Int = 0
}

/** ユーザー辞書へ登録する単語。既定値は JS 側で埋まっている。 */
class VoicevoxUserDictWordRecord : Record {
  @Field val surface: String = ""

  @Field val pronunciation: String = ""

  @Field val accentType: Int = 0

  @Field val wordType: String = "COMMON_NOUN"

  @Field val priority: Int = 5
}

/** voicevox-core 由来のエラーを JS へ伝えるための例外。 */
class VoicevoxException(message: String, cause: Throwable? = null) : CodedException(message, cause)

class ExpoVoicevoxModule : Module() {
  private val engine = VoicevoxEngine()

  /** voicevox-core の Synthesizer は同時実行できないので、重い処理はこのロックで直列化する。 */
  private val engineLock = Any()

  /** config plugin が配置したアセットの解決役。reactContext が要るので遅延生成する。 */
  private var assets: VoicevoxAssets? = null

  /**
   * 再生役。reactContext が要るので assets と同じく遅延生成する。
   *
   * `isSpeaking` が JS スレッドから engineLock を取らずに読むので `@Volatile`。
   * 生成は [playerInitLock] で守る（engineLock を使うと合成の完了まで待たされる）。
   */
  @Volatile private var player: VoicevoxPlayer? = null

  private val playerInitLock = Any()

  override fun definition() = ModuleDefinition {
    Name("ExpoVoicevox")

    Events("onPrepareProgress", "onSpeechStateChange")

    OnDestroy {
      // 停止は engineLock に依存しないので先に済ませる（合成の実行中でも即座に黙る）。
      player?.release()
      synchronized(engineLock) { engine.release() }
    }

    Function("getVersion") { GlobalInfo.getVersion() }

    Function("isInitialized") { engine.isInitialized }

    // 再生役がまだ無いなら鳴っているはずがない。ここで作ると reactContext 待ちで例外になる。
    Function("isSpeaking") { player?.isSpeaking ?: false }

    AsyncFunction("prepareAssets") {
      synchronized(engineLock) {
        val paths = prepareAssets()
        mapOf(
          "openJtalkDictDir" to paths.openJtalkDictDir,
          "voiceModelPaths" to paths.voiceModelPaths
        )
      }
    }

    AsyncFunction("initialize") { options: VoicevoxInitializeOptions ->
      synchronized(engineLock) {
        // 明示パスが両方そろっているときはアセットの準備を一切走らせない
        // （自前でモデルを管理している利用者に余計なダウンロードをさせないため）。
        val explicitDict = options.openJtalkDictDir
        val explicitModels = options.voiceModelPaths
        val dictDir: String
        val modelPaths: List<String>
        if (explicitDict != null && explicitModels != null) {
          dictDir = explicitDict
          modelPaths = explicitModels
        } else {
          val prepared = prepareAssets()
          dictDir = explicitDict ?: prepared.openJtalkDictDir
          modelPaths = explicitModels ?: prepared.voiceModelPaths
        }

        runWrappingErrors("failed to initialize voicevox-core") {
          engine.initialize(
            openJtalkDictDir = dictDir,
            voiceModelPaths = modelPaths,
            cpuNumThreads = options.cpuNumThreads
          )
        }
      }
    }

    AsyncFunction("getMetasJson") {
      synchronized(engineLock) {
        runWrappingErrors("failed to read the voice metadata") { engine.metasJson() }
      }
    }

    AsyncFunction("tts") {
      text: String,
      styleId: Int,
      enableInterrogativeUpspeak: Boolean,
      directory: String ->
      // 書き出し先の検証はロックの外で済ませる。
      val outputDir = outputDirectory(directory)
      synchronized(engineLock) {
        runWrappingErrors("speech synthesis failed") {
          writeWav(engine.tts(text, styleId, enableInterrogativeUpspeak), outputDir)
        }
      }
    }

    AsyncFunction("ttsFromKana") {
      kana: String,
      styleId: Int,
      enableInterrogativeUpspeak: Boolean,
      directory: String ->
      val outputDir = outputDirectory(directory)
      synchronized(engineLock) {
        runWrappingErrors("speech synthesis failed") {
          writeWav(engine.ttsFromKana(kana, styleId, enableInterrogativeUpspeak), outputDir)
        }
      }
    }

    AsyncFunction("createAudioQueryJson") { text: String, styleId: Int ->
      synchronized(engineLock) {
        runWrappingErrors("failed to create the audio query") {
          engine.createAudioQueryJson(text, styleId)
        }
      }
    }

    AsyncFunction("createAudioQueryFromKanaJson") { kana: String, styleId: Int ->
      synchronized(engineLock) {
        runWrappingErrors("failed to create the audio query") {
          engine.createAudioQueryFromKanaJson(kana, styleId)
        }
      }
    }

    AsyncFunction("createAccentPhrasesJson") { text: String, styleId: Int ->
      synchronized(engineLock) {
        runWrappingErrors("failed to create the accent phrases") {
          engine.createAccentPhrasesJson(text, styleId)
        }
      }
    }

    AsyncFunction("createAccentPhrasesFromKanaJson") { kana: String, styleId: Int ->
      synchronized(engineLock) {
        runWrappingErrors("failed to create the accent phrases") {
          engine.createAccentPhrasesFromKanaJson(kana, styleId)
        }
      }
    }

    AsyncFunction("replaceMoraDataJson") { accentPhrasesJson: String, styleId: Int ->
      synchronized(engineLock) {
        runWrappingErrors("failed to replace the mora data") {
          engine.replaceMoraDataJson(accentPhrasesJson, styleId)
        }
      }
    }

    AsyncFunction("replacePhonemeLengthJson") { accentPhrasesJson: String, styleId: Int ->
      synchronized(engineLock) {
        runWrappingErrors("failed to replace the phoneme length") {
          engine.replacePhonemeLengthJson(accentPhrasesJson, styleId)
        }
      }
    }

    AsyncFunction("replaceMoraPitchJson") { accentPhrasesJson: String, styleId: Int ->
      synchronized(engineLock) {
        runWrappingErrors("failed to replace the mora pitch") {
          engine.replaceMoraPitchJson(accentPhrasesJson, styleId)
        }
      }
    }

    // 推論を伴わないので直列化する必要が無い。
    AsyncFunction("audioQueryFromAccentPhrasesJson") { accentPhrasesJson: String ->
      runWrappingErrors("failed to build the audio query") {
        engine.audioQueryFromAccentPhrasesJson(accentPhrasesJson)
      }
    }

    AsyncFunction("synthesis") {
      audioQueryJson: String,
      styleId: Int,
      enableInterrogativeUpspeak: Boolean,
      directory: String ->
      val outputDir = outputDirectory(directory)
      synchronized(engineLock) {
        runWrappingErrors("speech synthesis failed") {
          writeWav(engine.synthesis(audioQueryJson, styleId, enableInterrogativeUpspeak), outputDir)
        }
      }
    }

    AsyncFunction("speak") {
      text: String,
      styleId: Int,
      enableInterrogativeUpspeak: Boolean,
      audioSession: String,
      promise: Promise ->
      speakWav(audioSession, promise) { engine.tts(text, styleId, enableInterrogativeUpspeak) }
    }

    AsyncFunction("speakFromKana") {
      kana: String,
      styleId: Int,
      enableInterrogativeUpspeak: Boolean,
      audioSession: String,
      promise: Promise ->
      speakWav(audioSession, promise) {
        engine.ttsFromKana(kana, styleId, enableInterrogativeUpspeak)
      }
    }

    AsyncFunction("speakFromAudioQuery") {
      audioQueryJson: String,
      styleId: Int,
      enableInterrogativeUpspeak: Boolean,
      audioSession: String,
      promise: Promise ->
      speakWav(audioSession, promise) {
        engine.synthesis(audioQueryJson, styleId, enableInterrogativeUpspeak)
      }
    }

    // 合成の実行中でも即座に止められるよう、engineLock は取らない。
    AsyncFunction("stopSpeaking") { requirePlayer().stop() }

    AsyncFunction("setUserDictWords") { words: List<VoicevoxUserDictWordRecord> ->
      val converted = words.map { toVoicevoxWord(it) }
      synchronized(engineLock) {
        runWrappingErrors("failed to update the user dictionary") {
          engine.setUserDictWords(converted)
        }
      }
    }

    AsyncFunction("loadUserDictFile") { path: String ->
      synchronized(engineLock) {
        runWrappingErrors("failed to load the user dictionary") { engine.loadUserDictFile(path) }
      }
    }

    AsyncFunction("saveUserDictFile") { path: String ->
      synchronized(engineLock) {
        runWrappingErrors("failed to save the user dictionary") { engine.saveUserDictFile(path) }
      }
    }

    AsyncFunction("finalize") { synchronized(engineLock) { engine.release() } }
  }

  private fun toVoicevoxWord(record: VoicevoxUserDictWordRecord): VoicevoxWord {
    // UserDictWord.Type は enum ではないので when で引く。
    val wordType =
      when (record.wordType) {
        "PROPER_NOUN" -> UserDictWord.Type.PROPER_NOUN
        "COMMON_NOUN" -> UserDictWord.Type.COMMON_NOUN
        "VERB" -> UserDictWord.Type.VERB
        "ADJECTIVE" -> UserDictWord.Type.ADJECTIVE
        "SUFFIX" -> UserDictWord.Type.SUFFIX
        else ->
          throw VoicevoxException("unknown user dictionary word type: ${record.wordType}")
      }
    return VoicevoxWord(
      surface = record.surface,
      pronunciation = record.pronunciation,
      accentType = record.accentType,
      wordType = wordType,
      priority = record.priority
    )
  }

  /** config plugin が配置したアセットを使える状態にする。進捗は JS へイベントで流す。 */
  private fun prepareAssets(): VoicevoxAssetPaths {
    val context =
      appContext.reactContext
        ?: throw VoicevoxException("the Android context is not available yet")
    val resolver = assets ?: VoicevoxAssets(context).also { assets = it }
    return runWrappingErrors("failed to prepare the voicevox assets") {
      resolver.prepare { progress -> sendEvent("onPrepareProgress", progress.toEventMap()) }
    }
  }

  /**
   * voicevox-core の型付き例外をそのまま JS へ流すと種類が多すぎるので、
   * メッセージを保ったまま 1 つの CodedException にまとめる。
   */
  private inline fun <T> runWrappingErrors(context: String, block: () -> T): T {
    try {
      return block()
    } catch (error: CodedException) {
      throw error
    } catch (error: Throwable) {
      val detail = error.message ?: error::class.java.simpleName
      throw VoicevoxException("$context: $detail", error)
    }
  }

  /**
   * 再生役を用意する。reactContext が要るので遅延生成する。
   *
   * AsyncFunction は同時に走りうるので、二重生成しないようロックの中で作る。取り逃すと
   * 誰も参照していない AudioTrack が鳴り続けて止められなくなる。
   */
  private fun requirePlayer(): VoicevoxPlayer =
    synchronized(playerInitLock) {
      player?.let {
        return it
      }
      val context =
        appContext.reactContext
          ?: throw VoicevoxException("the Android context is not available yet")
      val created = VoicevoxPlayer(context)
      created.onStateChange = { id, state, reason ->
        sendEvent(
          "onSpeechStateChange",
          mapOf("id" to id, "state" to state.jsValue, "reason" to reason)
        )
      }
      player = created
      created
    }

  /**
   * 再生系に共通する「engineLock の中で合成し、再生はロックの外で始める」流れ。
   *
   * 再生の完了をロックの中で待つと、鳴っているあいだ次の合成を始められない。
   * `VoicevoxPlayer.play` は専用スレッドを起こして即座に戻る。
   */
  private fun speakWav(audioSession: String, promise: Promise, synthesize: () -> ByteArray) {
    try {
      val mode =
        VoicevoxAudioSessionMode.parse(audioSession)
          ?: throw VoicevoxException("unknown audio session mode: $audioSession")
      val player = requirePlayer()
      val id = player.begin()
      val wav =
        try {
          synchronized(engineLock) {
            runWrappingErrors("speech synthesis failed") { synthesize() }
          }
        } catch (error: Throwable) {
          // 合成に失敗したら自分の予約だけ畳む。鳴っている音は止めない。
          player.cancel(id)
          throw error
        }
      val result = runWrappingErrors("playback failed") { player.play(id, wav, mode) }
      promise.resolve(result.toResultMap())
    } catch (error: CodedException) {
      promise.reject(error)
    } catch (error: Throwable) {
      promise.reject(
        VoicevoxException(error.message ?: error::class.java.simpleName, error)
      )
    }
  }

  /** 書き出し先の指定を実ディレクトリへ変換する。 */
  private fun outputDirectory(directory: String): File {
    val context =
      appContext.reactContext
        ?: throw VoicevoxException("the Android context is not available yet")
    return when (directory) {
      "cache" -> context.cacheDir
      // filesDir は Android Auto Backup（上限 25MB）の対象。WAV を貯め込むと上限に当たる。
      "document" -> context.filesDir
      else -> throw VoicevoxException("unknown output directory: $directory")
    }
  }

  /** 合成結果を指定のディレクトリへ書き出し、そのパスを返す。 */
  private fun writeWav(wav: ByteArray, baseDir: File): String {
    val outputDir = File(baseDir, "expo-voicevox")
    if (!outputDir.exists() && !outputDir.mkdirs()) {
      throw VoicevoxException("could not create the output directory: ${outputDir.absolutePath}")
    }
    val outputFile = File(outputDir, "${UUID.randomUUID()}.wav")
    outputFile.writeBytes(wav)
    return outputFile.absolutePath
  }
}
